src/gridfs.rs

```diff
@@ -0,0 +1,299 @@
+pub mod options;
+
+use core::task::{Context, Poll};
+use std::{io, pin::Pin};
+
+use crate::{
+    concern::{ReadConcern, WriteConcern},
+    cursor::Cursor,
+    error::{Error, Result},
+    selection_criteria::SelectionCriteria,
+    Database,
+};
+
+use bson::{oid::ObjectId, Bson, DateTime, Document};
+use futures_util;
+use options::*;
+use serde::{Deserialize, Serialize};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+pub const DEFAULT_BUCKET_NAME: &'static str = "fs";
+pub const DEFAULT_CHUNK_SIZE_BYTES: u32 = 255 * 1024;
+
+// Contained in a "chunks" collection for each user file
+struct Chunk {
+    id: ObjectId,
+    files_id: Bson,
+    n: u32,
+    // default size is 255 KiB
+    data: Vec<u8>,
+}
+
+/// A collection in which information about stored files is stored. There will be one files
+/// collection document per stored file.
+#[derive(Serialize, Deserialize)]
+pub struct FilesCollectionDocument {
+    id: Bson,
+    length: i64,
+    chunk_size: u32,
+    upload_date: DateTime,
+    filename: String,
+    metadata: Document,
+}
+
+/// Struct for storing GridFS managed files within a [`Database`].
+pub struct GridFsBucket {
+    // Contains a "chunks" collection
+    pub(crate) db: Database,
+    pub(crate) options: GridFsBucketOptions,
+}
+
+// TODO: RUST-1399 Add documentation and example code for this struct.
+pub struct GridFsUploadStream {
+    files_id: Bson,
+}
+
+impl GridFsUploadStream {
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.files_id
+    }
+
+    /// Consumes the stream and uploads data in the stream to the server.
+    pub async fn finish(self) {
+        todo!()
+    }
+
+    /// Aborts the upload and discards the upload stream.
+    pub async fn abort(self) {
+        todo!()
+    }
+}
+
+impl tokio::io::AsyncWrite for GridFsUploadStream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<tokio::io::Result<usize>> {
+        todo!()
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+}
+
+pub struct GridFsDownloadStream {
+    files_id: Bson,
+}
+
+impl GridFsDownloadStream {
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.files_id
+    }
+}
+
+impl tokio::io::AsyncRead for GridFsDownloadStream {
```

Comment:
we'll also want to implement `futures` version of `AsyncRead` here. Users can decide which trait to import, so there won't be any clashing issues.